### PR TITLE
network: address: drop deprecated temporary address

### LINKED ISSUE #19838: After an undefined amount of time, IPv6PrivacyExtensions=yes is no longer honored
**systemd version the issue has been seen with**
 > systemd 248 (248.3-2-arch)
> +PAM +AUDIT -SELINUX -APPARMOR -IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY +P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified

**Used distribution**
 > archlinux

**Linux kernel version used** (`uname -a`)

 > `Linux toxoplasmosis 5.12.4-2-ck-skylake #1 SMP PREEMPT Sun, 16 May 2021 02:40:56 +0000 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 > `IPv6PrivacyExtensions=yes` is permanently honored. The SLAAC IPv6 is *never* the IP that the system uses to reach the internet; `curl -6 icanhazip.com` *never* returns the `scope global dynamic mngtmpaddr noprefixroute` IP.

**Unexpected behaviour you saw**
 > `curl -6 icanhazip.com` returns my SLAAC IPv6 address, which does not honor `IPv6PrivacyExtensions=yes`

**Steps to reproduce the problem**
* `IPv6PrivacyExtensions=yes` in the adequate systemd.network(5) file.
* Wait a long time (21 days in my case; previous times was about 24 hours)
* `curl -6 icanhazip.com`

**Additional program output to the terminal or log subsystem illustrating the issue**

```
% networkctl 
IDX LINK      TYPE      OPERATIONAL SETUP
[...]
  2 wlp3s0    wlan      routable    configured
[...]
```
```
% ip addr show wlp3s0 
2: wlp3s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 11:22:33:44:55:66 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.11/24 brd 192.168.1.255 scope global dynamic wlp3s0
       valid_lft 79566sec preferred_lft 79566sec
    inet6 home:prefix:b287:b87d:c905:6312/64 scope global temporary deprecated dynamic 
       valid_lft 1770sec preferred_lft 0sec
    inet6 home:prefix:ab82:f20c:d84f:b6d/64 scope global temporary deprecated dynamic 
       valid_lft 1770sec preferred_lft 0sec
    inet6 home:prefix:1122:33ff:fe44:5566/64 scope global dynamic mngtmpaddr noprefixroute 
       valid_lft 1770sec preferred_lft 570sec
    inet6 fe80::1122:33ff:fe44:5566/64 scope link 
       valid_lft forever preferred_lft forever
```
```
% sysctl net.ipv6.conf.wlp3s0
net.ipv6.conf.wlp3s0.accept_dad = 1
net.ipv6.conf.wlp3s0.accept_ra = 0
net.ipv6.conf.wlp3s0.accept_ra_defrtr = 1
net.ipv6.conf.wlp3s0.accept_ra_from_local = 0
net.ipv6.conf.wlp3s0.accept_ra_min_hop_limit = 1
net.ipv6.conf.wlp3s0.accept_ra_mtu = 1
net.ipv6.conf.wlp3s0.accept_ra_pinfo = 1
net.ipv6.conf.wlp3s0.accept_ra_rt_info_max_plen = 0
net.ipv6.conf.wlp3s0.accept_ra_rt_info_min_plen = 0
net.ipv6.conf.wlp3s0.accept_ra_rtr_pref = 1
net.ipv6.conf.wlp3s0.accept_redirects = 1
net.ipv6.conf.wlp3s0.accept_source_route = 0
net.ipv6.conf.wlp3s0.addr_gen_mode = 0
net.ipv6.conf.wlp3s0.autoconf = 1
net.ipv6.conf.wlp3s0.dad_transmits = 1
net.ipv6.conf.wlp3s0.disable_ipv6 = 0
net.ipv6.conf.wlp3s0.disable_policy = 0
net.ipv6.conf.wlp3s0.drop_unicast_in_l2_multicast = 0
net.ipv6.conf.wlp3s0.drop_unsolicited_na = 0
net.ipv6.conf.wlp3s0.enhanced_dad = 1
net.ipv6.conf.wlp3s0.force_mld_version = 0
net.ipv6.conf.wlp3s0.force_tllao = 0
net.ipv6.conf.wlp3s0.forwarding = 0
net.ipv6.conf.wlp3s0.hop_limit = 64
net.ipv6.conf.wlp3s0.ignore_routes_with_linkdown = 0
net.ipv6.conf.wlp3s0.keep_addr_on_down = 0
net.ipv6.conf.wlp3s0.max_addresses = 16
net.ipv6.conf.wlp3s0.max_desync_factor = 600
net.ipv6.conf.wlp3s0.mc_forwarding = 0
net.ipv6.conf.wlp3s0.mldv1_unsolicited_report_interval = 10000
net.ipv6.conf.wlp3s0.mldv2_unsolicited_report_interval = 1000
net.ipv6.conf.wlp3s0.mtu = 1500
net.ipv6.conf.wlp3s0.ndisc_notify = 0
net.ipv6.conf.wlp3s0.ndisc_tclass = 0
net.ipv6.conf.wlp3s0.optimistic_dad = 0
net.ipv6.conf.wlp3s0.proxy_ndp = 0
net.ipv6.conf.wlp3s0.ra_defrtr_metric = 1024
net.ipv6.conf.wlp3s0.regen_max_retry = 3
net.ipv6.conf.wlp3s0.router_probe_interval = 60
net.ipv6.conf.wlp3s0.router_solicitation_delay = 1
net.ipv6.conf.wlp3s0.router_solicitation_interval = 4
net.ipv6.conf.wlp3s0.router_solicitation_max_interval = 3600
net.ipv6.conf.wlp3s0.router_solicitations = -1
net.ipv6.conf.wlp3s0.rpl_seg_enabled = 0
net.ipv6.conf.wlp3s0.seg6_enabled = 0
net.ipv6.conf.wlp3s0.seg6_require_hmac = 0
sysctl: permission denied on key 'net.ipv6.conf.wlp3s0.stable_secret'
net.ipv6.conf.wlp3s0.suppress_frag_ndisc = 1
net.ipv6.conf.wlp3s0.temp_prefered_lft = 86400
net.ipv6.conf.wlp3s0.temp_valid_lft = 604800
net.ipv6.conf.wlp3s0.use_oif_addrs_only = 0
net.ipv6.conf.wlp3s0.use_optimistic = 0
net.ipv6.conf.wlp3s0.use_tempaddr = 2 # <----- Most important
```
```INI
/etc/systemd/network/20-wireless.network
[Match]
MACAddress=11:22:33:44:55:66

[Network]
DHCP=yes
IPv6PrivacyExtensions=yes
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #19838.

### CODE REVIEW

File: src/network/networkd-address.c [hunk 1]
```diff
@@ -1394,6 +1394,20 @@ int manager_rtnl_process_address(sd_netlink *rtnl, sd_netlink_message *message,
                         address->scope = tmp->scope;
                         address_set_lifetime(address, &cinfo);
                         address_enter_configured(address);
+
+                        if (address->family == AF_INET6 &&
+                            address->source == NETWORK_CONFIG_SOURCE_FOREIGN &&
+                            FLAGS_SET(address->flags, IFA_F_SECONDARY | IFA_F_DEPRECATED)) {
+
+                                /* The temporary address becomes deprecated, i.e. preferred lifetime == 0. */
```
Comment (bluca) [inline src/network/networkd-address.c]: /becomes/became/

**Review by bluca — APPROVED** [review]

**Comment by yuwata** [comment]
@bluca Thank you for the review. The above point is addressed. Upgrading the green label.

**Comment by bluca** [comment]
> @bluca Thank you for the review. The above point is addressed. Upgrading the green label.

I set the reporter label as I thought we might want to wait for the users who opened the issue to confirm it fixes it? Or is it something worth having independently of that?

**Comment by Jookia** [comment]
Just to note in this PR I'll be testing this over the next few days.

**Comment by Jookia** [comment]
I'm not too familiar with Linux's IP stack, but will this drop addresses that are in use? Say I have a TCP connection open using a temporary address just as it gets removed. Would that fail it?

**Comment by Jookia** [comment]
I'm not too familiar with Linux's IP stack, but will this drop addresses that are in use? Say I have a TCP connection open using a temporary address just as it gets removed. Would that fail it?

**Comment by yuwata** [comment]
Hmm, at least, we need to remove deprecated address(es) at a reasonable timing. Otherwise, the kernel keeps many deprecated addresses, and refuses to generate a new temporary address.

Actually, the current logic may be too aggressive... Maybe, wait for at least one active temporary address become available?

**Comment by yuwata** [comment]
(Please do not merge this. I will reconsider the issue.)

**Comment by yuwata** [comment]
Hmm, after re-reading kernel's `addrconf_verify_rtnl()` in net/ipv6/addrconf.c, now I think the issue is caused by a race in the kernel.

`addrconf_verify_rtnl()` does the following:
1. if the preferred lifetime of an address is zero, then make the address deprecated.
2. if the preferred lifetime of a temporary address is positive, but very small (defaults to less than 5sec), then create a new temporary address, if the new address generation is not triggered on the address.
3. re-trigger `addrconf_verify_rtnl()` after the minimum preferred lifetime (for non-temporary addresses) or preferred lifetime - 5sec (for temporary addresses). But it cannot be earlier than 1sec after the function ends.

So, if `addrconf_verify_rtnl()` is once called when the preferred lifetime of a temporary address is slightly above the threshold, and the function takes so long (>~ 4sec), then on the next call, the temporary address enters deprecated without generating a new temporary address.

Several points I have not understand yet:
- One additional concern is that `ipv6_create_tempaddr()` is called with `bool block = true` in `addrconf_verify_rtnl()`. I've not investigated what the `block` argument does. But it sounds dangerous...
- Also, I am not sure how kernel accurately trigger the delayed tasks.

I am not 100% sure about the investigation, but still I hope that dropping the deprecated address(es) fixes the issue. At least, it can be a workaround for the issue. Why? When there exists no temporary address corresponding to a public address, then updating the address will generate a new temporary address.

**Comment by keszybz** [comment]
OK, let's merge this. We'll have rc1 released soon, and we should probably mention this in the notes so people can watch out for regressions.

**Comment by Jookia** [comment]
Please, please, PLEASE revert. This kills connections that use the addresses.

**Comment by yuwata** [comment]
Yeah, will do.

**Pull request merged**